src/core/shared_memory_manager.cc

```diff
@@ -117,34 +151,73 @@ SharedMemoryManager::RegisterSharedMemory(
   }
 
   // register (or re-register)
-  void* mapped_addr;
-  int shm_fd = -1;
-
-  // don't re-open if shared memory is already open
-  for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
-       ++itr) {
-    if (itr->second->shm_key_ == shm_key) {
-      shm_fd = itr->second->shm_fd_;
-      break;
+  if (kind == 0) {
+    void* mapped_addr;
+    int shm_fd = -1;
+
+    // don't re-open if shared memory is already open
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == shm_key) {
+        shm_fd = itr->second->shm_fd_;
+        break;
+      }
     }
-  }
 
-  // open and set new shm_fd if new shared memory key
-  if (shm_fd == -1) {
-    RETURN_IF_ERROR(OpenSharedMemoryRegion(shm_key, &shm_fd));
-  }
+    // open and set new shm_fd if new shared memory key
+    if (shm_fd == -1) {
+      RETURN_IF_ERROR(OpenSharedMemoryRegion(shm_key, &shm_fd));
+    }
+
+    Status status = MapSharedMemory(shm_fd, offset, byte_size, &mapped_addr);
+    if (!status.IsOk()) {
+      return Status(
+          RequestStatusCode::INVALID_ARG,
+          "failed to register shared memory region '" + name + "'");
+    }
+    shared_memory_map_.insert(std::make_pair(
+        name, std::unique_ptr<SharedMemoryInfo>(new SharedMemoryInfo(
+                  name, shm_key, offset, byte_size, shm_fd, mapped_addr,
+                  nullptr, kind, device_id))));
+  } else {
+#ifdef TRTIS_ENABLE_GPU
+    void *mapped_addr = nullptr, *cuda_ipc_addr;
+    size_t shm_byte_size = 0, shm_offset = 0;
+
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == shm_key) {
+        mapped_addr = itr->second->mapped_addr_;
+        shm_byte_size = itr->second->byte_size_;
+        shm_offset = itr->second->offset_;
+        break;
+      }
+    }
+
+    if (shm_byte_size < sizeof(ipcCUDA_t)) {
+      return Status(
+          RequestStatusCode::INVALID_ARG,
+          "shared memory region '" + name +
+              "' must be at least of the size of cuda IPC representation");
+    }
 
-  Status status = MapSharedMemory(shm_fd, offset, byte_size, &mapped_addr);
-  if (!status.IsOk()) {
+    // Get CUDA shared memory base address
+    void* shm_mapped_addr = (void*)((uint8_t*)mapped_addr + shm_offset);
+    ipcCUDA_t* shm_cuda_rep = reinterpret_cast<ipcCUDA_t*>(shm_mapped_addr);
+    OpenCUDAIPCRegion(shm_cuda_rep, &cuda_ipc_addr);
+
+    shared_memory_map_.insert(std::make_pair(
+        name, std::unique_ptr<SharedMemoryInfo>(new SharedMemoryInfo(
+                  name, shm_key, offset, byte_size, -1, nullptr, cuda_ipc_addr,
+                  kind, device_id))));
+#else
     return Status(
         RequestStatusCode::INVALID_ARG,
-        "failed to register shared memory region '" + name + "'");
+        "Cannot register CUDA shared memory region '" + name +
```

Comment:
It is not that there is no valid GPU, it is that this build doesn't even support GPUs. So say "failed to register CUDA..., GPUs not supported"